### PR TITLE
3mf Import/Export

### PR DESCRIPTION
**Pull request by t-paul**
This adds import and export for the [3MF file format](http://www.3mf.io/) using [lib3mf](https://github.com/3MFConsortium/lib3mf).

See also #1331.

ToDo:
- [x] Mark as experimental feature
- [x] Build on Travis Ubuntu 12.04 and 14.04
- [x] Build on Travis OSX / homebrew (using openscad/homebrew-tap repo)
- [x] Mac release build (macosx_build_dependencies.sh)
- [x] Update MXE build (currently stuck, it builds on MXE but using the library crashes)
- [x] Add patch to MXE build ([MinGW issue](https://sourceforge.net/p/mingw-w64/bugs/565/) - wcstof crashes with endPtr == nullptr)
- [x] The build system should fail (or at least warn) if lib3mf wasn't found at qmake time.
- [x] Add "export-import" tests (see #1822), similar to how we handle other bidirectional formats, e.g. `add_cmdline_test(offpngtest EXE ${PYTHON_EXECUTABLE} SCRIPT ${CMAKE_SOURCE_DIR}/export_import_pngtest.py ARGS --openscad=${OPENSCAD_BINPATH} --format=OFF --render EXPECTEDDIR monotonepngtest SUFFIX png FILES ${EXPORT3D_TEST_FILES})`
- [x] Update macOS build script
- [x] Look into macOS install_name

### CODE REVIEW

**Comment by kintel** [comment]
The lib3mf build should work for Mac now. I wouldn't call what they've got a "build system", but it will work as long as they don't change it :/

**Comment by t-paul** [comment]
Yeah, it's just some generated file from an IDE or so. At least they accepted the minimal wrapper Makefile to allow installation (https://github.com/3MFConsortium/lib3mf/pull/9) so building on OBS, MXE and homebrew is ok-ish.

**Comment by t-paul** [comment]
Ok, Windows/MinGW issues are now fixed. The `wcstof()` bug in MinGW is fixed, but until this is going to find it's way into MXE, I've added a workaround to the `lib3mf` MXE build.
Export file creation needed a special case, to enforce binary stream format, otherwise `lib3mf` failed due to `CR` -> `CRLF` conversion.

Not sure yet, why the `precise` Travis build now fails to install dependencies.

**Comment by t-paul** [comment]
For some reason the `sudo` calls in the external `travis-before-install.sh` script did work for `trusty` but not in `precise`. I've changed config to call the whole script via `sudo`, which works for both.

I'm going to rebase this to get rid of the Travis testing stuff, so please update the branch if it's checked out somewhere...

**Comment by t-paul** [comment]
I'll have a look if it's possible to give some kind of final status output listing what optional features are enabled.

**Comment by t-paul** [comment]
Something like https://github.com/openscad/openscad/blob/inputdriver3/info.pri from the `inputdriver3` branch.

Used simply by adding `CONFIG+=info`

**Comment by kintel** [comment]
hm, I guess the challenge is that we're using qmake both to build dependable binaries as to build development binaries with potentially partial configs.

I'd say that we should make sure that the "dependable binaries" part is solid first.
..then make it more convenient to perform partial builds.

I guess there might be different requirements for various distro builds?

**Comment by rogersachan** [comment]
@t-paul Any update? Can you build a Win32 binary of your PR so I can test?

File: scripts/macosx-build-dependencies.sh [hunk 1]
```diff
@@ -53,6 +53,7 @@ PACKAGES=(
     "libxml2 2.9.7"
     "fontconfig 2.12.4"
     "hidapi 0.7.0"
+    "lib3mf a466df47231c02298dde295adb6075b0fc649eba"
```
Comment (kintel) [inline scripts/macosx-build-dependencies.sh]: I see there's a bunch of recent activity on lib3mf - should we bump this to a more recent version?

File: features/lib3mf.prf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# 3. system's standard include paths
+
+OPENSCAD_LIBRARIES_DIR = $$(OPENSCAD_LIBRARIES)
+
+!isEmpty(LIB3MF_INCLUDEPATH) {
+  LIB3MF_CFLAGS = -I$$LIB3MF_INCLUDEPATH -I$$LIB3MF_INCLUDEPATH/Model/COM
+} else {
+  !isEmpty(OPENSCAD_LIBRARIES_DIR) {
+    exists($$OPENSCAD_LIBRARIES_DIR/include/lib3mf/Model/COM/NMR_DLLInterfaces.h) {
+      LIB3MF_CFLAGS = -I$$OPENSCAD_LIBRARIES_DIR/include/lib3mf -I$$OPENSCAD_LIBRARIES_DIR/include/lib3mf/Model/COM
```
Comment (kintel) [inline features/lib3mf.prf]: It seems a bit odd to add some subpath under 3mf as our include path. Couldn't we just `#include <lib3mf/Model/COM/NMR_DLLInterfaces.h>` in our code?

File: scripts/macosx-build-dependencies.sh [hunk 3]
```diff
@@ -766,6 +767,25 @@ build_hidapi()
...
+  if [ ! -f $version.tar.gz ]; then
+    curl -LO https://github.com/3MFConsortium/lib3mf/archive/$version.tar.gz
+  fi
+  tar xzf $version.tar.gz
+  cd lib3mf-$version
+  CXXFLAGS="$CXXSTDFLAGS" make CC=cc CXX=c++ LD=c++ LDFLAGS="$LDSTDFLAGS" -j"$NUMCPU" -C Project/Lib3MFGCC
+  make -j"$NUMCPU" -C Project/Lib3MFGCC install INSTALL_ROOT=$DEPLOYDIR
+#  install_name_tool -id @rpath/libCGAL.dylib $DEPLOYDIR/lib/libCGAL.dylib
+#  install_name_tool -id @rpath/libCGAL_Core.dylib $DEPLOYDIR/lib/libCGAL_Core.dylib
+#  install_name_tool -change libCGAL.11.dylib @rpath/libCGAL.dylib $DEPLOYDIR/lib/libCGAL_Core.dylib
```
Comment (kintel) [inline scripts/macosx-build-dependencies.sh]: Note to self: Have a look at this

File: features/lib3mf.prf (hunk 2 above)
Comment (t-paul) [reply features/lib3mf.prf]: They had lots of internal references that need the multiple `-I` flags. Maybe that is fixed now and just the main `.../lib3mf` is enough.

Comment (t-paul) [reply features/lib3mf.prf]: Yep, that's fine now, I changed our code too, so the Model/COM deep link is not needed anymore.

**Comment by kintel** [comment]
@t-paul Looks like Info.plist.in is used by the cmake build system

**Comment by t-paul** [comment]
Then we always edited the wrong one `i.e. not the Info.plist.in` in the past?

**Comment by kintel** [comment]
I guess nobody builds binaries with cmake yet, so that file may be lagging behind.

**Pull request merged**